### ISSUE TITLE
Change to install pattern with zypper breaks SLES

### ISSUE THREAD

**Issue by lukeb2e** [dotless-de/vagrant-vbguest#219]
Hey,
the recent change by @vpereira to install dependencies on opensuse with patterns instead of with packages breaks support for SLES.

https://github.com/dotless-de/vagrant-vbguest/commit/7bf3b7dd0596a3119e18055dae29973f0cb2e0d6

Testing confirms that after reverting the change, everything works as intended.

Tested on: SLES 12 / 11.4

**Comment by vpereira** [comment]
Hey @glookie1, 
so either we do a fine tuning on supported SUSE Linux versions or it won't be able to support both, I think,  right?

**Comment by glookie1** [comment]
Well I don't know if there are too many downsides to installing the required packages instead of the patterns on OpenSuse. If OpenSuse requires the patterns, then yes, it won't be able to support both.

An additional check for OpenSuse vs. SLES would probably be preferred, but I don't know how you would differentiate between OpenSuse and SLES as the currently checked values are identical.

Validating the information written in /etc/os-release might be the safest way, but feels hacky.
This is the recommended way at least for SLES12/SLES11.4.

```
NAME="SLES"
VERSION="12-SP1"
VERSION_ID="12.1"
PRETTY_NAME="SUSE Linux Enterprise Server 12 SP1"
ID="sles"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:suse:sles:12:sp1"
```

The os-release file for SLES 11.4 is identical ( except for Version numbers ;) )

**Comment by vpereira** [comment]
Yes opensuse is done via patterns. 

Maybe the solution would be the installer suse and the class method ````match?```` should implement a logic like https://github.com/dotless-de/vagrant-vbguest/blob/d36ba2e4ae00577b4169bf1336a6a3264277c1e7/lib/vagrant-vbguest/installers/oracle.rb

**Comment by vpereira** [comment]
I will try to implement it today.

**Comment by fnordfish** [comment]
Hi there,

thanks for debugging this. The installer could read some file contents from the guest machine containing the exact os version and make some switches on that.

It's a bit tricky, but basically works like this: https://github.com/dotless-de/vagrant-vbguest/blob/d36ba2e4ae00577b4169bf1336a6a3264277c1e7/lib/vagrant-vbguest/installers/base.rb#L133

**Comment by glookie1** [comment]
/etc/os-release exists since SLES 11.4 - only previous versions use /etc/SuSE-release

```
SUSE Linux Enterprise Server 11 (x86_64)
VERSION = 11
PATCHLEVEL = 4
# This file is deprecated and will be removed in a future service pack or release.
# Please check /etc/os-release for details about this release.
```

Therefore testing for SuSE-release might be a problem.

Something akin to this might be enough, if you don't require backwards compatibility to older SUSE Versions:
grep "SLES" /etc/os-release

**Comment by vpereira** [comment]
Hi @glookie1 

can you try the version of my repository? 

here is my commit https://github.com/vpereira/vagrant-vbguest/commit/199b15d0efb49e4b927344d673d9895276409a81 

if it works for you, I can open a pull request.

**Comment by vpereira** [comment]
To really fix it, Maybe we should go one level up into the Vagrant::Machine, Vagrant::Systems::Base, or other related classes.

**Comment by glookie1** [comment]
Currently it is not executing the new code in suse.rb and still executing opensuse.rb. Adding a check which will always fail tells me that it seems that it doesn't load the suse.rb. 
```
The guest's platform ("suse") is currently not supported, will try generic Linux method...
```
I'm not sure wether that would help, because I don't think that there is a SLES System Base in Virtualbox.

**Comment by vpereira** [comment]
Maybe it's related with the Installer priority. I will check it.

**Comment by glookie1** [comment]
I think the self.sles test isn't working. Adding the following line for debug reasons, I get the following output:
```
printf "self.distro(vm): %s has_zypper?(vm): %s sles?(vm): %s\n", self.distro(vm).inspect, has_zypper?(vm).inspect, sles?(vm).inspect
```
```
self.distro(vm): :suse has_zypper?(vm): true sles?(vm): false
```

**Comment by vpereira** [comment]
ok, funny that I tested in a SLE-11-SP4 the commands:

````
grep ID=\"sles\" /etc/os-release
````

and

````
which zypper
````
and it returned 0 as expected.

can you check if you get 0 after you execute the grep command? 

I'm commuting that's why I wasn't able to test it myself .. as soon as I'm at office, I can test it.

**Comment by glookie1** [comment]
For testing I disabled the sles check and added a similar check to opensuse. After trying again, it seems that the script itself is working, the problem seems to be that the communicate_to(vm).test was not supposed to be used in the way we do here...
Later today I should have time again to debug the issue.
```
sles?(vm): false ## <-- this is probably the issue
```
Is the vm.communicate something provided by Vagrant?

Just tested it, both commands return 0. Somewhere something strange happens...

**Comment by vpereira** [comment]
maybe do a git pull, because I did an update, in my first commit I forgot to require suse.rb..

**Comment by glookie1** [comment]
Latest Version with me still tries to use opensuse installer for sles.
As far as I can tell: 
* OpenSuSE Installer is now executed/called after SuSE Installer -> fixed
* `communicate_to(vm).test "grep ID=\"sles\" /etc/os-release"` always returns false
I tried playing around with it, but all I figured out to do was to get it to always return true, which isn't better in any way.
Seems like somewhere on the way it is doing something weird with the return variable. Currently I am unsure wether the problem is in the vbguest plugin or in vagrant itself...

I used the following code to Debug - once I am at home I will be able to try properly. Sadly I cant get git to work with our corporate proxy...
```
def self.sles?(vm)
  test = communicate_to(vm).test "grep ID=\"sles\" /etc/os-release"
  printf "SLES Debug: %s/%s", test.inspect, test.to_s
  return test
end
```
